### ISSUE TITLE
definition file for "do" operator doesn't match the do(null, fn)

### ISSUE THREAD

**Issue by Davilink** [ReactiveX/rxjs#3528]
**RxJS version:**
5.5.6
**Code to reproduce:**
````js
    new Observable((observer) => {
      observer.next("Yeah !");
      observer.error("error !");
    }).do(null, err => {
      console.log('oops!!!!');
    }).subscribe(something => console.log(something));
````
**Expected behavior:**
Able to use the do(null, fn);

**Actual behavior:**
I can't compile my application.
I have to use a dummy function to make it work.

````js
    new Observable((observer) => {
      observer.next("Yeah !");
      observer.error("error !");
    }).do(() => {}, err => {
      console.log('oops!!!!');
    }).subscribe(something => console.log(something));
````

**Additional information:**
RxJS 4 had the doOnError, but RxJS 5 merged that operator with the normal " do", with passing a « null » value has the first arguments, but it doesn't works.

**Comment by benlesh** [comment]
I'm unable to reproduce your error in JavaScript: http://jsbin.com/kewohibice/edit?js,console

I suspect the issue you're hitting is in TypeScript, in that case you can pass `undefined` and you'll be fine.

**Better**

I highly suggest you use:  `do({ error(err) { console.log('oops' } })` which is more readable and explicit.

**Comment by Davilink** [comment]
The problem is indeed with Typescript.

![image](https://user-images.githubusercontent.com/1498932/38703727-099ee4a6-3e72-11e8-86ef-08d55da6b3f4.png)

Because in the do.d.ts

![image](https://user-images.githubusercontent.com/1498932/38703696-f4863c90-3e71-11e8-861a-c57fd16384f3.png)

So we cannot use null neither undefined.

It would be needed to have something like:

![image](https://user-images.githubusercontent.com/1498932/38703900-8b4ef7ac-3e72-11e8-9e99-43f78c11bfb3.png)

**Comment by cartant** [comment]
It's not so much a problem with TypeScript, as it's a problem with the type definitions. The same happens with `tap`.

This is a minor annoyance that I've come across, too. And I've got around it using `!`, like this:

```ts
tap(undefined!, undefined!, () => console.log("complete")
```
`null!` would also work.

Basically, a signature like this:

```ts
export function tap<T>(
  next?: (x: T) => void,
  error?: (e: any) => void,
  complete?: () => void
): MonoTypeOperatorFunction<T>;
```

says that `next` is optional, but if an argument is passed, it **must be** `(x: T) => void`. So if passing `undefined` is acceptable usage, the signature should be:

```ts
export function tap<T>(
  next?: ((x: T) => void) | undefined,
  error?: ((e: any) => void) | undefined,
  complete?: (() => void) | undefined
): MonoTypeOperatorFunction<T>;
```

And if `null` is also acceptable, it should be in the union, too.

**Comment by Davilink** [comment]
>  next?: ((x: T) => void) | undefined
undefined is not needed here, because it's implicit by the use of « ? » (`next?` vs `next`)

I didn't thought about using the non-null assertion operator.

**Comment by cartant** [comment]
@Davilink Yeah. You are right. However, I'm now at a complete loss when it comes to explaining why passing an `undefined` was effecting TypeScript errors for me only a few weeks ago. Neither the typings for `do` nor `tap` have changed in months. I guess something was fixed in TypeScript 2.8.

Anyway, after reading Ben's comment, using a partial observer looks like a better way to go.
